### PR TITLE
[FLINK-24939][table] Support `SHOW CREATE CATALOG` syntax

### PR DESCRIPTION
**Pull request by liyubin117**
## What is the purpose of the change

Support `SHOW CREATE CATALOG` syntax to output the catalog name and relevant properties, which allows you to easily reuse the created catalogs.

## Brief change log

* SHOW CREATE CATALOG catalog_name

## Verifying this change

This change added tests and can be verified as follows:

flink-table/flink-sql-client/src/test/resources/sql/catalog_database.q
flink-table/flink-sql-gateway/src/test/resources/sql/catalog_database.q

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? yes

### CODE REVIEW

**Comment by liyubin117** [comment]
@wuchong Hi, CI passed, Could you please take a review? thanks very much :)

**Comment by LadyForest** [comment]
> @LadyForest @xuyangzhong @hackergin Hi, CI passed, Could you please take a review? thanks very much :)

Thanks for reaching out and I will take a look ASAP

File: docs/content.zh/docs/dev/table/sql/show.md [hunk 1]
```diff
@@ -102,6 +103,22 @@ tEnv.executeSql("SHOW CURRENT CATALOG").print();
 // |      default_catalog |
 // +----------------------+
 
+// create a catalog
+tEnv.executeSql("CREATE CATALOG cat2 WITH (...)")
```
Comment (LadyForest) [inline docs/content.zh/docs/dev/table/sql/show.md:107]:
```suggestion
tEnv.executeSql("CREATE CATALOG cat2 WITH (...)");
```

File: docs/content.zh/docs/dev/table/sql/show.md [hunk 2]
```diff
@@ -214,6 +231,22 @@ tEnv.executeSql("SHOW CATALOGS").print()
 // | default_catalog |
 // +-----------------+
 
+// create a catalog
+tEnv.executeSql("CREATE CATALOG cat2 WITH (...)")
+
+// show create catalog
+tEnv.executeSql("SHOW CREATE CATALOG cat2").print();
```
Comment (LadyForest) [inline docs/content.zh/docs/dev/table/sql/show.md:238]:
```suggestion
tEnv.executeSql("SHOW CREATE CATALOG cat2").print()
```

File: docs/content.zh/docs/dev/table/sql/show.md [hunk 3]
```diff
@@ -316,6 +349,22 @@ table_env.execute_sql("SHOW CATALOGS").print()
 # | default_catalog |
 # +-----------------+
 
+# create a catalog
+table_env.execute_sql("CREATE CATALOG cat2 WITH (...)")
+
+# show create catalog
+table_env.execute_sql("SHOW CREATE CATALOG cat2").print();
```
Comment (LadyForest) [inline docs/content.zh/docs/dev/table/sql/show.md:356]:
```suggestion
table_env.execute_sql("SHOW CREATE CATALOG cat2").print()
```

File: docs/content.zh/docs/dev/table/sql/show.md [hunk 4]
```diff
@@ -411,6 +460,14 @@ table_env.execute_sql("SHOW FULL MODULES").print()
...
 default_catalog
 
+Flink SQL> create catalog cat2 WITH (...);
+[INFO] Execute statement succeeded.
+ 
+Flink SQL> show create catalog cat2;
+CREATE CATALOG `cat2` WITH (
+  ...
+)
+
```
Comment (LadyForest) [inline docs/content.zh/docs/dev/table/sql/show.md:470]: Nit: let key word be capitalized to keep consistent with other examples.

File: docs/content/docs/dev/table/sql/show.md [hunk 5]
```diff
@@ -102,6 +103,22 @@ tEnv.executeSql("SHOW CURRENT CATALOG").print();
 // |      default_catalog |
 // +----------------------+
 
+// create a catalog
+tEnv.executeSql("CREATE CATALOG cat2 WITH (...)")
```
Comment (LadyForest) [inline docs/content/docs/dev/table/sql/show.md:107]: ditto

File: docs/content/docs/dev/table/sql/show.md [hunk 6]
```diff
@@ -214,6 +231,22 @@ tEnv.executeSql("SHOW CATALOGS").print()
 // | default_catalog |
 // +-----------------+
 
+// create a catalog
+tEnv.executeSql("CREATE CATALOG cat2 WITH (...)")
+
+// show create catalog
+tEnv.executeSql("SHOW CREATE CATALOG cat2").print();
```
Comment (LadyForest) [inline docs/content/docs/dev/table/sql/show.md:238]: ditto

File: docs/content/docs/dev/table/sql/show.md [hunk 7]
```diff
@@ -316,6 +349,22 @@ table_env.execute_sql("SHOW CATALOGS").print()
 # | default_catalog |
 # +-----------------+
 
+# create a catalog
+table_env.execute_sql("CREATE CATALOG cat2 WITH (...)")
+
+# show create catalog
+table_env.execute_sql("SHOW CREATE CATALOG cat2").print();
```
Comment (LadyForest) [inline docs/content/docs/dev/table/sql/show.md:356]: ditto

File: docs/content/docs/dev/table/sql/show.md [hunk 8]
```diff
@@ -411,6 +460,14 @@ table_env.execute_sql("SHOW FULL MODULES").print()
 Flink SQL> SHOW CATALOGS;
 default_catalog
 
+Flink SQL> create catalog cat2 WITH (...);
```
Comment (LadyForest) [inline docs/content/docs/dev/table/sql/show.md:463]: ditto

File: docs/content/docs/dev/table/sql/show.md [hunk 9]
```diff
@@ -504,6 +561,38 @@ SHOW CURRENT CATALOG
...
 
+## SHOW CREATE CATALOG
+
+```sql
+SHOW CREATE CATALOG catalog_name
+```
+
+Show creation statement for an existing catalog.
+
+The output includes the catalog's name and relevant properties, which allows you to easily reuse the created catalogs.
```
Comment (LadyForest) [inline docs/content/docs/dev/table/sql/show.md:572]: Nit: I think the phrase "easily resuse the created catalogs" doesn't quite capture the intended meaning. A more accurate description would be "gain an intuitive understanding of the underlying catalog's metadata."

**Review by LadyForest — CHANGES_REQUESTED** [review]
@liyubin117 Thanks for the contribution! The changes look good to me in general and I just left some minor comments. PTAL

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowCreateCatalog.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+import java.util.Collections;
+import java.util.List;
+
+/** SHOW CREATE CATALOG sql call. */
+public class SqlShowCreateCatalog extends SqlShowCreate {
+
+    public static final SqlSpecialOperator OPERATOR =
+            new SqlSpecialOperator("SHOW CREATE CATALOG", SqlKind.OTHER_DDL);
+
+    protected final SqlIdentifier catalogName;
```
[the remark this answers is not in the document]
Comment (liyubin117) [reply flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowCreateCatalog.java:38]: Yes, After this SqlNode is converted into `SqlShowCreateCatalogOperation`, this variable `catalogName` is needed when calling `ctx.getCatalogManager().getCatalogDescriptor()`.

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowCreateCatalog.java [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    @Override
+    public void unparse(SqlWriter writer, int leftPrec, int rightPrec) {
+        writer.keyword("SHOW CREATE CATALOG");
+        sqlIdentifier.unparse(writer, leftPrec, rightPrec);
+    }
+
+    public String catalogName() {
+        return catalogName.getSimple();
```
[the remark this answers is not in the document]
Comment (liyubin117) [reply flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowCreateCatalog.java:66]: Do you mean replace the instance name? I found that `SqlCreateCatalog` also use `catalogName`, WDYT?
```
public class SqlCreateCatalog extends SqlCreate {

    public static final SqlSpecialOperator OPERATOR =
            new SqlSpecialOperator("CREATE CATALOG", SqlKind.OTHER_DDL);

    private final SqlIdentifier catalogName;

    private final SqlNodeList propertyList;
```

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowCreateCatalog.java [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+    protected final SqlIdentifier catalogName;
+
+    public SqlShowCreateCatalog(SqlParserPos pos, SqlIdentifier catalogName) {
+        super(pos, catalogName);
+        this.catalogName = catalogName;
+    }
+
+    public SqlIdentifier getCatalogName() {
+        return catalogName;
```
[the remark this answers is not in the document]
Comment (liyubin117) [reply flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowCreateCatalog.java:46]: ditto

File: docs/content.zh/docs/dev/table/sql/show.md (hunk 1 above)
Comment (liyubin117) [reply docs/content.zh/docs/dev/table/sql/show.md]: Thank you for pointing it out :)

File: docs/content.zh/docs/dev/table/sql/show.md (hunk 2 above)
Comment (liyubin117) [reply docs/content.zh/docs/dev/table/sql/show.md]: done

File: docs/content.zh/docs/dev/table/sql/show.md (hunk 3 above)
Comment (liyubin117) [reply docs/content.zh/docs/dev/table/sql/show.md]: done

File: docs/content.zh/docs/dev/table/sql/show.md (hunk 4 above)
Comment (liyubin117) [reply docs/content.zh/docs/dev/table/sql/show.md]: done

File: docs/content/docs/dev/table/sql/show.md (hunk 5 above)
Comment (liyubin117) [reply docs/content/docs/dev/table/sql/show.md]: done

File: docs/content/docs/dev/table/sql/show.md (hunk 6 above)
Comment (liyubin117) [reply docs/content/docs/dev/table/sql/show.md]: done

File: docs/content/docs/dev/table/sql/show.md (hunk 7 above)
Comment (liyubin117) [reply docs/content/docs/dev/table/sql/show.md]: done

File: docs/content/docs/dev/table/sql/show.md (hunk 8 above)
Comment (liyubin117) [reply docs/content/docs/dev/table/sql/show.md]: done

File: docs/content/docs/dev/table/sql/show.md (hunk 9 above)
Comment (liyubin117) [reply docs/content/docs/dev/table/sql/show.md]: Thank you for pointing it out, I have updated as you said :)

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowCreateCatalog.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (liyubin117) [reply flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowCreateCatalog.java:46]: ditto

**Comment by liyubin117** [comment]
@LadyForest Thanks for your suggestions, I have updated as you said and CI passed, PTAL :)

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowCreateCatalog.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (LadyForest) [reply flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowCreateCatalog.java:38]: I'm still confused about this. The base class `SqlShowCreate` already defined a protected variable called `sqlIdentifier`. Why not use it to represent the catalog name, just like other subclasses(`SqlShowCreateView`, `SqlShowCreateTable`) do?

**Review by LadyForest — COMMENTED** [review]
I'm still confused about this. The base class `SqlShowCreate` already defined a protected variable called `sqlIdentifier`. Why not use it to represent the catalog name, just like other subclasses(`SqlShowCreateView`, `SqlShowCreateTable`) do?

**Comment by liyubin117** [comment]
> I'm still confused about this. The base class `SqlShowCreate` already defined a protected variable called `sqlIdentifier`. Why not use it to represent the catalog name, just like other subclasses(`SqlShowCreateView`, `SqlShowCreateTable`) do?

I did not notice that sqlIdentifier has been defined, it is my fault, thanks for pointing it out 👍

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowCreateCatalog.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (liyubin117) [reply flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowCreateCatalog.java]: done :)

**Review by LadyForest — APPROVED** [review]
Thanks for the update, LGTM

**Pull request merged**